polygon/bridge/bridge.go

```diff
@@ -0,0 +1,200 @@
+	// create new db
+	db, err := node.OpenDatabase(ctx, config, kv.PolygonBridgeDB, name, readonly, logger)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Bridge{
+		DB:                       db,
+		log:                      logger,
+		borConfig:                borConfig,
+		fetchSyncEvents:          fetchSyncEvents,
+		eventMap:                 map[uint64]IDRange{},
+		lastProcessedBlockNumber: 0,
+		lastProcessedEventID:     0,
+		stateClientAddress:       libcommon.HexToAddress(borConfig.StateReceiverContract),
+	}, nil
+}
+
+func (b *Bridge) Run(ctx context.Context) error {
+	// start syncing
+	b.log.Debug(bridgeLogPrefix("Bridge is running"))
+
+	// get last known sync ID
+	lastEventID, err := GetLatestEventID(b.DB)
+	if err != nil {
+		return err
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		// get all events from last sync ID to now
+		to := time.Now()
+		events, err := b.fetchSyncEvents(ctx, lastEventID+1, to, 0)
+		if err != nil {
+			return err
+		}
+
+		if len(events) != 0 {
+			b.ready = false
+			if err := AddEvents(b.DB, events); err != nil {
+				return err
+			}
+
+			lastEventID = events[len(events)-1].ID
+		} else {
+			b.ready = true
+			time.Sleep(30 * time.Second)
+		}
+
+		b.log.Debug(bridgeLogPrefix(fmt.Sprintf("got %v new events, last event ID: %v, ready: %v", len(events), lastEventID, b.ready)))
+	}
+}
+
+func (b *Bridge) Close() {
+	b.DB.Close()
+}
+
+// EngineService interface implementations
+
+// ProcessNewBlocks iterates through all blocks and constructs a map from block number to sync events
+func (b *Bridge) ProcessNewBlocks(ctx context.Context, blocks []*types.Block) error {
+	for _, block := range blocks {
+		// check if block is start of span
+		if b.isSprintStart(block.NumberU64()) {
+			continue
+		}
+
+		t := time.Unix(int64(block.Time()), 0)
+		lastDBID, err := GetLastSpanEventID(b.DB, b.lastProcessedEventID, t)
+		if err != nil {
+			return err
+		}
+
+		if lastDBID != 0 && lastDBID > b.lastProcessedEventID {
+			b.log.Warn(bridgeLogPrefix(fmt.Sprintf("Creating map for block %d, start ID %d, end ID %d", block.NumberU64(), b.lastProcessedEventID, lastDBID)))
+			b.eventMap[block.NumberU64()] = IDRange{b.lastProcessedEventID, lastDBID}
+
+			b.lastProcessedEventID = lastDBID
+		}
+
+		b.lastProcessedBlockNumber = block.NumberU64()
+	}
+
+	return nil
+}
+
+// Synchronize blocks till bridge has map at tip
+func (b *Bridge) Synchronize(ctx context.Context, tip *types.Header) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+		if b.ready || b.lastProcessedBlockNumber >= tip.Number.Uint64() {
+			return nil
+		}
+	}
+}
+
+// Unwind deletes map entries till tip
+func (b *Bridge) Unwind(ctx context.Context, tip *types.Header) error {
+	for k := range b.eventMap {
+		if k <= tip.Number.Uint64() {
+			delete(b.eventMap, k)
```

Comment:
fyi Unwind may need to delete the data in the DB too - but let's leave this for another time since we are not dealing with Unwinds yet